### PR TITLE
Android logging

### LINKED ISSUE #1199: Android logging
## Is your feature request related to a problem? Please describe 👀

Am wondering if we are able to log / view logs in android?
The f/e is logging via bugsnag already - am more concerned about the server logs. 

I'd like to know if the server logs somewhere on android ( I think it doesn't ), and if the somewhere is retrievable, somehow ( I think it isn't ).

## Describe the solution you'd like 🎁

The log settings should be set to log to file, on android. 
File location will be in app data, so we will also need a way to export or email the log somewhere. Alternatively the log could be viewable through the app, though I'm reluctant as that is more work.

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #1199 

# 👩🏻‍💻 What does this PR do? 
 
Android now implements `fast-log` including the file rotation strategy. The log level is set to `info` and the log directory is the file directory of the app.
It has meant removing `android_logger` which connects the rust logging to android logcat - it isn't possible to have two logging subsystems in use.

Here's how it looks:

Screenshot 2023-03-02 at 9 00 17 AM

https://user-images.githubusercontent.com/9192912/222265559-ce146fea-0928-4198-9198-db0a023049a8.mp4

I'd used an `ArrayList` when reading in the log file and tried to send an array to the client app.. but couldn't get that working. It came through as a string.. so I've reverted back to just sending a string to the client. Not ideal, but enough for now. I've kept the display as a big block of string, as this allows 'select all' and 'copy n paste' which may be helpful until we have the option of exporting or emailing the log file

The log viewer functionality is only for android clients currently - In a server environment we can view the actual logs, and the desktop client will be connecting to a server so won't need to view the logs (maybe?)

There's a lot more that we can do - the immediate need was to have some way of viewing what is going on in order to assist support, and I think this is enough.

Further possibilities:
- enable the configuration of the log level
- allow selection of the log file to be viewed ( currently it only shows the active log, not previous ones )
- add an option to export the log to a text file
- add an option to email the log file ( to support? )

# 🧪 How has/should this change been tested? 

Have installed on the galaxy tablet and tested manually.
You can run the usual commands to build and test if you like:

```
yarn android:build:debug
adb install -r ./packages/android/app/build/outputs/apk/debug/open-msupply-1.1.3-debug.apk

```

## 📃 Documentation

- [ ] Could add to the android docs.. but probably not for external users

### CODE REVIEW

**Review by andreievg — COMMENTED** [review]
I am thinking of a use case where error occurs before user can get to the admin page (and there could be a lot of use cases for that), if we log to a file inside `app` folder, these files are not accessible outside of the app in release version (in debug version they are accessible). This would mean we would loose the logging we would otherwise had (accessed via log cat). 

I am finding it hard to find the issue where suggestion was made to use logcat reader, wanted to know what the response for that issue was. Also was wondering if MEMDM has this built in, ability to examine logs remotely ?

File: server/android/src/android.rs [hunk 1]
```diff
@@ -64,9 +60,15 @@ pub mod android {
...
             // sync settings need to be configured at runtime
             sync: None,
-            logging: None,
+            logging: Some(
+                LoggingSettings::new(LogMode::File, service::settings::Level::Info)
+                    .with_directory(files_dir.to_string_lossy().to_string()),
+            ),
         };
 
+        logging_init(settings.logging.clone());
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/android/src/android.rs:69]: Would it be useful to have file logging + Android log output? e.g. if we can't get to the settings UI?  Then you don't have to manually look at the file on the device...

Think you could call the [AndroidLogger](https://docs.rs/android_logger/latest/android_logger/struct.AndroidLogger.html#) like this:
https://github.com/rbatis/fast_log#custom-logimpl-do_log-method

[the remark this answers is not in the document]
Comment (mark-prins) [reply server/android/src/android.rs:69]: thanks, yes. will do.

**Comment by mark-prins** [comment]
> if MEMDM has this built in, ability to examine logs remotely ?

I found [this page](https://www.manageengine.com/mobile-device-management/how-to/generate-adb-logs.html) which implies that it isn't possible, otherwise why would you use adb & USB debugging?

Then [this one](https://www.manageengine.com/mobile-device-management/how-to/logs-how-to.html#mdm_android_logs_from_server) about sending logs to ME support, which is for MEMDM logs rather than logcat output.

I think if I can save them out to the `/sdcard` instead of the app folder then we can at least access them?
The logcat is great, but also very noisy - I think if we have the OMS logs separate it would be easier to see what's going on

**Comment by mark-prins** [comment]
I've popped the logs into the `Documents` folder on the device in db3bdf097

Which means that we can access the files to view / email like so:

Screenshot 2023-03-03 at 6 04 02 PM

**Review by andreievg — APPROVED** [review]
Looks good @mark-prins, thanks for the changes. 

Would suggest having an issue to improve on the logging window (it's technically not needed now, but if there is a way to access log in UI):
* I think it's a good idea to have the functionality available for all types of omSupply (would help support team, they'll just need to access it one way)
* Send logs via email 
* Adding this way of accessing API to desktop flavour would mean API
* As you've pointed out pagination etc..

It sounds like it's adding up, maybe just opening the file in text editor is the best way to work it ¯\_(ツ)_/¯

**Comment by mark-prins** [comment]
> Would suggest having an issue to improve on the logging window

Thanks - added 1277 as a low priority issue. Agree that it's a bit of work and isn't strictly necessary as we can use external tools to do most of the things. Would be great to have in-app though if we do get the space to work on it!

**Pull request merged**